docs/contributing/sources-and-providers.md

```diff
@@ -0,0 +1,53 @@
+# Sources and Providers
+
+ExternalDNS supports swapping out endpoint **sources** and DNS **providers** and both sides are pluggable. There currently exist three sources and four provider implementations.
+
+### Sources
+
+Sources are an abstraction over any kind of source of desired Endpoints, e.g.:
+* a list of Service objects from Kubernetes
+* a random list for testing purposes
+* an aggregated list of multiple nested sources
+
+The `Source` interface has a single method called `Endpoints` that should return all desired Endpoint objects as a flat list.
+
+```go
+type Source interface {
+	Endpoints() ([]*endpoint.Endpoint, error)
+}
+```
+
+All sources live in package `source`.
+
+* `ServiceSource`: collects all Services of `type=LoadBalancer` that have an external IP and returns them as Endpoint objects. The desired DNS name corresponds to an annotation set on the Service or is compiled from the Service attributes via the FQDN Go template string.
```

Comment:
nitpick: `type=LoadBalancer` is not correct as far as I remember, we do not filter on type.